docs/design/storage_classes.md

```diff
@@ -0,0 +1,413 @@
+2.  Use arbitrary labels and selectors to drive PV binding claims
+
+```go
+// PersistentVolumeClaimSpec describes the common attributes of storage devices
+// and allows a Source for provider-specific attributes
+type PersistentVolumeClaimSpec struct {
+    // Contains the types of access modes required
+    AccessModes []PersistentVolumeAccessMode `json:"accessModes,omitempty"`
+    // PersistentVolumeSelector is a selector which must be true for the claim to bind to a volume
+    PersistentVolumeSelector *unversioned.LabelSelector `json:"persistentVolumeSelector,omitempty"`
+    // Resources represents the minimum resources required
+    Resources ResourceRequirements `json:"resources,omitempty"`
+    // VolumeName is the binding reference to the PersistentVolume backing this claim
+    VolumeName string `json:"volumeName,omitempty"`
+}
+```
+
+### New `StorageClass` resource
+
+A new API extension resource should be added called `StorageClass`:
+
+```go
+// Storage class is an admin-defined
+type StorageClass struct {
+    unversioned.TypeMeta `json:",inline"`
+    ObjectMeta           `json:"metadata,omitempty"`
+
+    Description string `json:"description,omitempty"`
+}
+```
+
+It should be noted here that `StorageClass` in this proposal is basically
+meant only to allow administrators to record a model of their taxonomy of
+storage classes within Kubernetes.  Future features such as dynamic
+provisioning will extend this type with additional fields.
+
+The usual implementation tasks will exist for this new type:
+
+1.  Add to extensions API
+2.  Add REST storage implementation
+3.  Add `kubectl` bits like describe
+
+The labels of the `StorageClass` describe how a volume belonging to that
+storage class should be labeled.
+
+For example, say that an administrator defines a `StorageClass` called `my-
+storage-class` described by the labels `type=bar` and `color=foo`:
+
+```yaml
+apiVersion: extensions/v1
+kind: StorageClass
+metadata:
+  name: my-storage-class
+  labels:
+    type: bar
+    color: foo
+description: A very special storage class
+```
+
+#### New API resource or ConfigMap?
+
+During discussion of this topic in
+[kubernetes/17056](https://github.com/kubernetes/kubernetes/pull/17056), it
+was assumed that we would use `ConfigMap` to hold information about storage
+classes.  This would not be impossible to do, but there are several challenges
+due to the fact that `PersistentVolume` is a global resource and `ConfigMap`
+is namespaced.   We would need a virtual resource which appeared to be global
+but which was a view on a namespaced resource -- no such pattern is currently
+implemented anywhere in the system.  For that virtual resource, we would need
+to solve all the permission/policy issues in Kubernetes and downstream
+projects like OpenShift.  We would also likely need to define reserved key
+names for certain attributes of `StorageClass` within the `ConfigMap` `data`
```

Comment:
@bgrant0607 I agree that it has almost no meaning in the context of this proposal -- however, it will have a very well defined meaning and widespread appeal (my subjective opinion here) once we add dynamic provisioning on top of it.